### PR TITLE
chore: take advantage of CheckpointContext handling sharded data [MLG-13]

### PR DESCRIPTION
**Pull request by drh-determined-ai**
## Description

Since `CheckpointContext` now supports sharding, the workloads loop in `DeepSpeedTrialController` does not need to collect data spread across nodes.

## Test Plan

Run some Deepspeed examples and note that checkpoints are being created as expected (same files of approximately the same sizes as before these changes). This includes the saving and loading of checkpoints when experiments are paused and unpaused.

The following examples and configurations thereof ran without errors and without new warnings:
- `cifar10_cpu_offloading`
  - `zero_3_cpu_offload.yaml`
- `cifar10_moe`
  - `zero_stages.yaml` with ZeRO stage 3
- `gpt_neox`
   - `19M.yml`
   - `global_num_gpus/pp_size/mp_size = 1` 
   - `global_num_gpus/pp_size/mp_size = 2`

Dynamic agent instances were of type `g4dn.4xlarge` or `g4dn.12xlarge`.
 
## Commentary (optional)

The above experiments were run; it was confirmed that the checkpoints were created as expected (with the correct files and file sizes).

## Checklist

- [x] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Comment by aciborowska** [comment]
Saving checkpoints looks good to me. I would be more convinced if we added tests to inspect checkpoint content and ensure all the files are there.

We may also want to add a selector to restore_path so that each node gets only relevant files, as pointed out by Liam here:
```
    def _load(self, load_path: pathlib.Path) -> None:
        # Right now we will load all checkpoint shards on each node regardless of which
        # checkpoints are needed.
        # TODO (Liam): revisit later to optimize sharded checkpoint loading.
```
@liamcli iirc looking at filenames should be sufficient or do we need something more advanced?

**Comment by drh-determined-ai** [comment]
Please comment on the log messages I added. I'm certain they can be improved: what information is too extra, what's missing, is the wording accurate to what is actually happening?

File: harness/determined/core/_checkpoint.py [hunk 1]
```diff
@@ -260,6 +260,7 @@ def _upload_single(
         *,
         selector: Optional[Callable[[str], bool]] = None,
     ) -> str:
+        logger.info(f"Uploading checkpoint to {ckpt_dir} metadata={metadata}")
```
Comment (aciborowska) [inline harness/determined/core/_checkpoint.py:263]: I would keep all the new logs at the debug level.
Also, I think this log is incorrect: we are uploading `ckpt_dir` content to the storage (not checkpoint to `ckpt_dir`). Similar problem in the next log.

**Review by aciborowska — APPROVED** [review]
nitpick on logging and lgtm!

**Pull request merged**